app/src/main/java/de/rochefort/childmonitor/VolumeView.java

```diff
@@ -49,37 +44,25 @@ public VolumeView(Context context, @Nullable AttributeSet attrs, int defStyleAtt
     }
 
     private void init() {
-        volume = 0;
-        maxVolume = 0.25;
         paint = new Paint();
-        volumeHistory = new LinkedList<>();
         paint.setColor(Color.rgb(255, 127, 0));
     }
 
-    public void onAudioData(short[] data) {
-        double sum = 0;
-        for (int i = 0; i < data.length; i++) {
-            double rel = data[i] / ((double)128);
-            sum += Math.pow(rel, 2);
-        }
-        volume = sum / data.length;
-        if (volume > maxVolume) {
-            maxVolume = volume;
-        }
-        volumeHistory.addLast(volume);
-        while (volumeHistory.size() > MAX_HISTORY) {
-            volumeHistory.removeFirst();
-        }
-        invalidate();
-    }
-
     @Override
     protected void onDraw(Canvas canvas) {
-        int height = canvas.getHeight();
-        int width = canvas.getWidth();
-        double relativeBrightness = 0;
-        double normalizedVolume = volume / maxVolume;
-        relativeBrightness = Math.max(0.3, normalizedVolume);
+        final VolumeHistory volumeHistory = _volumeHistory;
+        if (volumeHistory == null) {
+            return;
+        }
+
+        final int height = canvas.getHeight();
+        final int width = canvas.getWidth();
+
+        final VolumeHistory history = volumeHistory.getSnapshot(width);
```

Comment:
Just noticed that your changes to the graph drawing broke the scrolling once the full canvas width is reached.
Not sure what the purpose of these changes are?
Also note that the snapshotting is no longer needed as we now ensure that the volume history is written to on the main event loop, so reading from it within onDraw is thread safe now.

Reply:
> Not sure what the purpose of these changes are?

It moves more logic out of the loop.

Since `volumeHistory.size() <= width`:
-  `leftMost = 0`.
- and we can iterate until `volumeHistory.size()`

```
            if (i == leftMost (== 0)) {
                xPrev = xNext;
            } 
```
Can be moved out of the loop, if we start with `1`, and initialize the `x/yPrev` values with the ones for `0`, and copy them from the `x/yNext` at the end of the loop.
We loose by that drawing a line for `(i=0)` to `(i=0)`

Finally: `/ maxVolume * (graphHeight) -> * graphHeight / maxVolume -> * grapScale`
 
I think, the bug is actually [here](https://github.com/enguerrand/child-monitor/pull/9/files#diff-34a7e83689bb93f2560f06c0fa9615b983b70c576eaded6d86313f08e89e473cR67-R69). That needs to be done the way you did the `leftMost`logic in the painter.

Reply:
It also used the wrong history instance further down, but I fixed that too now.

Reply:
I couldn't resist to remove the snapshot stuff. Sorry for committing one file at a time, I can't convince the Github UI to change two files in one commit ^^